### ISSUE TITLE
Improving signal management for EDS quantification

### ISSUE THREAD

**Issue by CorentinLG** [hyperspy/hyperspy#1780]
The four issues below are related to the post-modelling treatment of EDS data, for quantification purposes. 

1) after a model has been fitted, we use "result = m.get_lines_intensity()" to extract the peak heights. However, this function returns a list (ex : result[0] is the signal for the first x-ray line). For further quantification operation, it may be better to have a dictionnary instead. One could manipulate "result" as one single objet and call it for each specific xray_line.

2) As for "m.get_lines_intensity()", the "s.quantification()" function returns a list whereas it would be probably more convenient if it returned a dictionnary node (within the model ?).

3) "m.get_line_intensity()" returns the intensity each line separately (say Fe_Ka). However, Fe_Kb is also measured and fitted, and is always proportional to Fe_Ka. I think that there should be an option for "m.get_lines_intensity()" to return (Fe_Ka + Fe_Kb), since it is the value that is used for quantification afterwards.

4) To use k- or zeta-factors for quantification, "s.quantification() takes "result" and "kfactors" as arguments. The k-factors list must have the same length as the "result" list, which length is the number of considered x-ray lines. In practice, one has to order the kfactor list in the same fashion as the result list. This is annoying and leads to systematic mistakes. K-factors being specific to each microscope, it would be convenient if one could enter them within the (hyperspectral signal?) metadata, associated to their respective xray_lines. Then, s.quantification() would take only "result" as argument and respecting the ordering of each list would no longer be an issue.

**Comment by francisco-dlp** [comment]
I think that these are excellent usability enhancements that requires an API change. I've labelled it "next major".

**Comment by vidartf** [comment]
> Fe_Kb is also measured and fitted, and is always proportional to Fe_Ka

Note: This is not true if `m.free_sub_xray_lines_weight()` has been called.

**Comment by CorentinLG** [comment]
You are right. We need caution here. I think by default Ka and Kb should be summed directly, but we should keep our options open for when we want to work on each of them separately.

**Comment by k8macarthur** [comment]
I like the ideas in general. Personally I use my own dictionary to store all my EDX quantification library and quickly write that into a list. I'm not sure I quite understand how being able to store the cross-sections/z-factors etc into the metadata would speed things up. Otherwise being able to sum lines could be beneficial.

**Comment by CorentinLG** [comment]
Regarding storing k-factors : It would be easier if a given k-factor was directly associated to its x-ray-lines through the metadata/dictionnary. One would no longer have to sort out the different lists (x-ray-line intensities and k factors) for each specific case. Does that make sense?

**Comment by k8macarthur** [comment]
Ah got it. Yes this would simplify things a little.

**Comment by CorentinLG** [comment]
Is there a function at the moment that would return the statistical error of the EDS quantification? Theoretically, it is inversely proportional to the square root of the are of the corresponding gaussian. 
If not, I would suggest to include this calculation (and its result) as part of the s.quantification function. Then the result should maybe also be stored as part of the model or signal metadata ?

**Comment by k8macarthur** [comment]
I like the idea of trying to get error bars for the composition. One number won’t work though as it will vary across the image. Might be best to provide the relative rather than absolute error map so that people can include their errors from other places. E.g. in Constants like kfactors and probs current measurements.

Best wishes,

Katherine MacArthur

**Comment by CorentinLG** [comment]
Sure, we need an error map, which could be I think : the sq root of the gaussian area divided by the area itself multiplied  the quantified concentration, for each pixel. That would be the statistical counting error.

On top of that comes the errors on the calibration factors and/or probe current (for zeta factors). The two levels should be distinguished somehow...But I am not sure what s the best way to organize it.

**Comment by k8macarthur** [comment]
I've been thinking about this more recently. I think producing an error map with a get_error function which just produces the relative statistical counting error for each pixel. They people can simply multiply (or multiply the squares with any other errors they expect to have in their quantification?). Or they can add these relative errors as inputs to the get_error map.

**Comment by CorentinLG** [comment]
I agree.
This could be done as a specific get_error function as you suggest, or by passing an extra argument to the quantification function : s.quantification(calculate_error = True).

The other errors are specific to each method, so it might be difficult to integrate them in a method in a robust manner?

**Comment by k8macarthur** [comment]
Agreed, the other possible errors will vary so much that all we could realistically produce is the statistics error. For the remainder I would recommend making an example notebook which explains how to incorporate the remaining errors. If the output errors were all left as relative values it would be much easier for users to carry out addition of the squares later.

**Comment by k8macarthur** [comment]
In which case a simple function using the results from get_lines intensity should be able to determine the root N error from the maps. However, this will change depending on if more than one sub-shell is summed....
